core-rust/state-manager/src/store/parent_map.rs

```diff
@@ -0,0 +1,315 @@
+}
+
+impl NodeMetadataResolver {
+    /// Resolves the [`SubstateNodeMetadata`] for all newly-created Nodes found in the [`Create`]d
+    /// and [`Update`]d substates. (within the given [`SubstateChange`]s).
+    /// The resolution considers the trees (or rather: potentially-incomplete bottom-up tree
+    /// fragments) constructed within the given [`SubstateChange`]s, and the top fragments of these
+    /// trees already-existing in the given [`SubstateNodeMetadataStore`].
+    /// API note: the results are grouped by [`SubstateNodeMetadata`] (since it may be shared by
+    /// many [`NodeId`]s).
+    pub fn batch_resolve<S: SubstateNodeMetadataStore>(
+        metadata_store: &S,
+        substate_changes: &[SubstateChange],
+    ) -> impl Iterator<Item = (Vec<NodeId>, SubstateNodeMetadata)> {
+        // Index the new Nodes by their parent Substates (using `IndexedScryptoValue`):
+        let changed_nodes_by_parent = substate_changes
+            .iter()
+            .filter_map(|substate_change| match &substate_change.action {
+                ChangeAction::Create(bytes) | ChangeAction::Update(bytes) => {
+                    let owned_node_ids = IndexedScryptoValue::from_slice(bytes).unwrap().unpack().1;
+                    if owned_node_ids.is_empty() {
+                        return None;
+                    }
+                    let parent_reference = SubstateReference(
+                        substate_change.node_id,
+                        substate_change.partition_number,
+                        substate_change.substate_key.clone(),
+                    );
+                    Some((parent_reference, owned_node_ids))
+                }
+                ChangeAction::Delete => None,
+            })
+            .collect::<IndexMap<_, _>>();
+
+        // Invert the multimap (to obtain a Node -> parent Substate map):
+        let changed_parent_by_node = changed_nodes_by_parent
+            .iter()
+            .flat_map(|(parent, children)| children.iter().map(|child| (*child, parent.clone())))
+            .collect::<NonIterMap<_, _>>();
+
+        // Find the parents which do _not_ know their parent within the `SubstateChange`s:
+        let top_level_parent_node_ids = changed_nodes_by_parent
+            .keys()
+            .filter(|parent| !changed_parent_by_node.contains_key(&parent.0))
+            .map(|parent| parent.0)
+            .collect::<Vec<_>>();
+
+        // Load the missing metadata of these parents from the `SubstateNodeMetadataStore`:
+        let existing_metadatas = metadata_store.batch_get_metadata(&top_level_parent_node_ids);
+
+        // Construct a Node -> root map needed for the `NodeMetadataResolver` helper structure:
+        let existing_root_by_node = top_level_parent_node_ids
+            .into_iter()
+            .zip(existing_metadatas)
+            .filter_map(|(parent, existing_metadata)| {
+                existing_metadata.map(|metadata| (parent, metadata.root))
```

Comment:
If `existing_metadata` doesn't exist for `parent`, it could be because the `parent` is global. But it seems we filter that out instead of mapping to `(parent, parent)` - is this an issue? Or have I misunderstood something?

Reply:
Ahh this is handled by the resolver 👍  nice. ignore this.